File: src/main/webapp/js/service/claim-kpi-service.js
mainApp.service('claimKpiService', ['$http', '$q', '$log', 'GlobalHelper',  function($http, $q, $log, GlobalHelper){
	// Return public API.
    return({
    	findByJobNo:	findByJobNo,
    	saveByJobNo: 	saveByJobNo,
    	saveListByJobNo: saveListByJobNo,
    	deleteByJobNo:	deleteByJobNo,
    	getByPage:		getByPage,
    	getByJobNoYear:	getByJobNoYear
    });
	
    function findByJobNo(jobNo) {
		var request = $http({
			method: "get",
			url: "service/claimKpi/" + jobNo
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }

    function getByJobNoYear(jobNo, year, month) {
		var request = $http({
			method: "get",
			url: "service/claimKpi/getByYear/" + jobNo + "/" + year
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }
    
    function getByPage(page, size, direction, property,
    		noJob, 
			year, 
			month, 
			numberIssued, 
			amountIssued, 
			numberSubmitted, 
			amountSubmitted, 
			numberAssessed, 
			amountAssessed, 
			numberApplied, 
			amountApplied, 
			numberCertified, 
			amountCertified, 
			remarks,
		    eojSecured,
		    eojUnsecured,
		    eojTotal,
		    exceptionComment
		) {
		var request = $http({
			method: "get",
			url: "service/claimKpi/" + page + "/" + size,
			params: {
				direction: direction,
				property: property,
				noJob: noJob, 
				year: year, 
				month: month, 
				numberIssued: numberIssued, 
				amountIssued: amountIssued, 
				numberSubmitted: numberSubmitted, 
				amountSubmitted: amountSubmitted, 
				numberAssessed: numberAssessed, 
				amountAssessed: amountAssessed, 
				numberApplied: numberApplied, 
				amountApplied: amountApplied, 
				numberCertified: numberCertified, 
				amountCertified: amountCertified, 
				remarks: remarks,
				eojSecured: eojSecured,
				eojUnsecured: eojUnsecured,
				eojTotal: eojTotal,
				exceptionComment: exceptionComment
			}
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }
    
    function saveByJobNo(jobNo, kpi) {
		var request = $http({
			method: "post",
			url: "service/claimKpi/" + jobNo,
			data: kpi
		});
//		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
		return request;
    }
    
    function saveListByJobNo(jobNo, kpiList) {
		var request = $http({
			method: "post",
			url: "service/claimKpi/saveList/" + jobNo,
			data: kpiList
		});
//		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
		return request;
    }
    
    function deleteByJobNo(jobNo, id) {
		var request = $http({
			method: "delete",
			url: "service/claimKpi/" + jobNo + "/" + id
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }
}]);

File: src/main/java/com/gammon/pcms/service/ClaimKpiService.java
package com.gammon.pcms.service;

import com.gammon.pcms.dao.ClaimKpiRepository;
import com.gammon.pcms.model.ClaimKpi;
import edu.emory.mathcs.backport.java.util.Arrays;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Transactional
@Service
public class ClaimKpiService {

	private Logger logger = LoggerFactory.getLogger(getClass());
	
	@Autowired
	private ClaimKpiRepository repository;
	
	public List<ClaimKpi> findByJobNo(String jobNo) {
		return repository.findByNoJob(jobNo);
	}
	
	public List<ClaimKpi> getByYear(String jobNo, int year) {
		return repository.getByYear(jobNo, year);
	}
	
	public ClaimKpi save(String jobNo, ClaimKpi kpi) {
		ClaimKpi claimKpi = new ClaimKpi();
		try {
			kpi.setNoJob(jobNo);
			claimKpi = repository.save(kpi);
		} catch (Exception e) {
			e.printStackTrace();
		}
		return claimKpi;
	}
	
	public List<ClaimKpi> save(String jobNo, List<ClaimKpi> kpiList) {
		kpiList.forEach(kpi -> {
			kpi.setNoJob(jobNo);
		});
		return repository.save(kpiList);
	}
	
	public void delete(String jobNo, Long[] ids) {
		List<Long> idList = Arrays.asList(ids);
		idList.forEach(id -> {
			ClaimKpi kpiDb = repository.getOne((Long)id);
			if(kpiDb.getNoJob().equals(jobNo)) {
				repository.delete(kpiDb);
			} else {
				throw new IllegalAccessError("cannot delete claim kpi");
			}
		});
	}
	
	public Page<ClaimKpi> filterPagination(
			Pageable pageable,
			String noJob,
			String year, 
			String month, 
			String numberIssued, 
			String amountIssued, 
			String numberSubmitted,
			String amountSubmitted,
			String numberAssessed,
			String amountAssessed,
			String numberApplied,
			String amountApplied,
			String numberCertified,
			String amountCertified,
			String remarks,
			String eojSecured,
			String eojUnsecured,
			String eojTotal
			) {
		return repository.filterPagination(
				pageable, 
				noJob, 
				year  + "%", 
				month + "%", 
				numberIssued  + "%", 
				amountIssued  + "%", 
				numberSubmitted  + "%", 
				amountSubmitted  + "%", 
				numberAssessed  + "%", 
				amountAssessed  + "%", 
				numberApplied  + "%", 
				amountApplied  + "%", 
				numberCertified  + "%", 
				amountCertified  + "%", 
				remarks  + "%",
				eojSecured  + "%",
				eojUnsecured  + "%",
				eojTotal  + "%"
				);
	}
}
